engines/dfs.c

```diff
@@ -0,0 +1,564 @@
+		.help           = "List of pool replicated service ranks",
+		.category	= FIO_OPT_C_ENGINE,
+		.group		= FIO_OPT_G_DFS,
+	},
+#endif
+	{
+		.name           = NULL,
+	},
+};
+
+static int
+daos_fio_global_init(struct thread_data *td)
+{
+	struct daos_fio_options	*eo = td->eo;
+	uuid_t			pool_uuid, co_uuid;
+	daos_pool_info_t	pool_info;
+	daos_cont_info_t	co_info;
+	int			rc = 0;
+
+#if !defined(DAOS_API_VERSION_MAJOR) || DAOS_API_VERSION_MAJOR < 1
+	if (!eo->pool || !eo->cont || !eo->svcl) {
+#else
+	if (!eo->pool || !eo->cont) {
+		log_err("Missing required DAOS options\n");
+		return EINVAL;
+	}
+#endif
+
+	rc = daos_init();
+	if (rc != -DER_ALREADY && rc) {
+		log_err("Failed to initialize daos %d\n", rc);
+		td_verror(td, EINVAL, "daos_init");
+		return EINVAL;
+	}
+
+	rc = uuid_parse(eo->pool, pool_uuid);
+	if (rc) {
+		log_err("Failed to parse 'Pool uuid': %s\n", eo->pool);
+		td_verror(td, EINVAL, "uuid_parse(eo->pool)");
+		return EINVAL;
+	}
+
+	rc = uuid_parse(eo->cont, co_uuid);
+	if (rc) {
+		log_err("Failed to parse 'Cont uuid': %s\n", eo->cont);
+		td_verror(td, EINVAL, "uuid_parse(eo->cont)");
+		return EINVAL;
+	}
+
+	/** Connect to the DAOS pool */
+#if !defined(DAOS_API_VERSION_MAJOR) || DAOS_API_VERSION_MAJOR < 1
+	d_rank_list_t *svcl = NULL;
+
+	svcl = daos_rank_list_parse(eo->svcl, ":");
+	if (svcl == NULL) {
+		log_err("Failed to parse svcl\n");
+		td_verror(td, EINVAL, "daos_rank_list_parse");
+		return EINVAL;
+	}
+
+	rc = daos_pool_connect(pool_uuid, NULL, svcl, DAOS_PC_RW,
+			&poh, &pool_info, NULL);
+	d_rank_list_free(svcl);
+#else
+	rc = daos_pool_connect(pool_uuid, NULL, DAOS_PC_RW, &poh, &pool_info,
+			       NULL);
+#endif
+	if (rc) {
+		log_err("Failed to connect to pool %d\n", rc);
+		td_verror(td, EINVAL, "daos_pool_connect");
+		return EINVAL;
+	}
+
+	/** Open the DAOS container */
+	rc = daos_cont_open(poh, co_uuid, DAOS_COO_RW, &coh, &co_info, NULL);
+	if (rc) {
+		log_err("Failed to open container: %d\n", rc);
+		td_verror(td, EINVAL, "daos_cont_open");
+		(void) daos_pool_disconnect(poh, NULL);
+		return EINVAL;
+	}
+
+	/** Mount encapsulated filesystem */
+	rc = dfs_mount(poh, coh, O_RDWR, &dfs);
+	if (rc) {
+		log_err("Failed to mount DFS namespace: %d\n", rc);
+		td_verror(td, EINVAL, "dfs_mount");
+		(void) daos_pool_disconnect(poh, NULL);
+		(void) daos_cont_close(coh, NULL);
+		return EINVAL;
+	}
+
+	/** Retrieve object class to use */
+	if (eo->oclass) {
+		cid = daos_oclass_name2id(eo->oclass);
+	} else {
+		/** use default object class */
+		cid = OC_UNKNOWN;
+	}
+
+	return 0;
+}
+
+static void
+daos_fio_global_cleanup()
+{
+	int rc;
+
+	rc = dfs_umount(dfs);
+	if (rc)
+		log_err("failed to umount dfs: %d\n", rc);
+	rc = daos_cont_close(coh, NULL);
+	if (rc)
+		log_err("failed to close container: %d\n", rc);
+	rc = daos_pool_disconnect(poh, NULL);
+	if (rc)
+		log_err("failed to disconnect pool: %d\n", rc);
+	rc = daos_fini();
+	if (rc)
+		log_err("failed to finalize daos: %d\n", rc);
+}
```

Comment:
Don't you want to return/set an error so callers know something went wrong?

Reply:
Good point. fixed.